src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationReloadToken.cs

```diff
@@ -33,11 +33,60 @@ public class ConfigurationReloadToken : IChangeToken
         /// <param name="callback">The callback to invoke.</param>
         /// <param name="state">State to be passed into the callback.</param>
         /// <returns>The <see cref="CancellationToken"/> registration.</returns>
-        public IDisposable RegisterChangeCallback(Action<object?> callback, object? state) => _cts.Token.Register(callback, state);
+        public IDisposable RegisterChangeCallback(Action<object?> callback, object? state)
```

Comment:
Is it possible to refactor the duplicated code out into a shared helper?

https://github.com/dotnet/runtime/blob/e48558390c988a830040e355c55f729615c91246/src/libraries/Microsoft.Extensions.Primitives/src/CancellationChangeToken.cs#L32-L72

We typically do that by putting a file in https://github.com/dotnet/runtime/tree/main/src/libraries/Common/src/Extensions, and then "link" it into both projects.

See 

https://github.com/dotnet/runtime/blob/231255dac0fe015ebc6fb55466c75cf5b4e05f56/src/libraries/Microsoft.Extensions.Configuration.Binder/src/Microsoft.Extensions.Configuration.Binder.csproj#L13

and
https://github.com/dotnet/runtime/blob/199436b51fba6f0bee469c63ee42fa266d70222d/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Microsoft.Extensions.DependencyInjection.Abstractions.csproj#L15-L16

for an example.

Reply:
I've been exploring this, and the issue is that `RegisterChangeCallback` affects the ConfigurationChangeToken's `CancellationTokenSource` as well as its property `ActiveChangeCallbacks` which specifically cannot be passed as a ref or out.

Reply:
The common method could look something like:

```C#
internal static IDisposable UnsafeRegisterChangeCallback<T>(Action<object?> callback, object? state, CancellationToken token, Action<T> onFailure, T onFailureState)
{
#if NETCOREAPP || NETSTANDARD2_1
            try
            {
                return token.UnsafeRegister(callback, state);
            }
            catch (ObjectDisposedException)
            {
                // Reset the flag so that we can indicate to future callers that this wouldn't work.
                onFailure(onFailureState);
            }
#else
            // Don't capture the current ExecutionContext and its AsyncLocals onto the token registration causing them to live forever
            bool restoreFlow = false;
            if (!ExecutionContext.IsFlowSuppressed())
            {
                ExecutionContext.SuppressFlow();
                restoreFlow = true;
            }

            try
            {
                return token.Register(callback, state);
            }
            catch (ObjectDisposedException)
            {
                // Reset the flag so that we can indicate to future callers that this wouldn't work.
                onFailure(onFailureState);
            }
            finally
            {
                // Restore the current ExecutionContext
                if (restoreFlow)
                {
                    ExecutionContext.RestoreFlow();
                }
            }
#endif
            return NullDisposable.Instance;
}
```

And then then here we would call it like:

```C#
 public IDisposable RegisterChangeCallback(Action<object?> callback, object? state)
{
    UnsafeRegisterChangeCallback(
        callback,
        state,
        _cts.Token,
        static s => s.ActiveChangeCallbacks = false,  // Reset the flag so that we can indicate to future callers that this wouldn't work.
        this);
}
```